File: ios/AdyenDropInModule.swift
import Adyen

@objc(AdyenDropInModule)
class AdyenDropInModule: NSObject {
    internal var currentComponent: PresentableComponent?
    
    internal weak var presenter: UIViewController?
    
    private var apiClient: DefaultAPIClient?
    
    private var dropInConfiguration: DropInComponent.Configuration? = nil
    
    private var resolveCallback: RCTResponseSenderBlock? = nil
    
    private var rejectCallback: RCTResponseSenderBlock? = nil
    
    private var onSubmitCallback: RCTResponseSenderBlock? = nil
    
    private var onAdditionalDetailsCallback: RCTResponseSenderBlock? = nil
    
    // MARK: - RN module
    
    @objc
    static func requiresMainQueueSetup() -> Bool {
      return true
    }
    
    func resolve(_ arg: Any) {
        resolveCallback?.self([arg])
    }
    
    func reject(_ arg: Any) {
        rejectCallback?.self([arg])
    }
    
    @objc
    func setDropInConfig(_ config: NSDictionary?) {
        print("Called setDropInConfig")
        
        if (config == nil) {
            reject("setDropInConfig was called without a config object")
            return
        }
        
        guard let clientKey = config?["clientKey"] as? String,
              let environmentName = config?["environment"] as? String
        else {
            reject("setDropInConfig missing clientKey or environment")
            return
        }
        
        MemoryStorage.current.clientKey = clientKey
        
        if let shopperReference = config?["shopperReference"] as? String {
            MemoryStorage.current.shopperReference = shopperReference
        }
        
        if let countryCode = config?["countryCode"] as? String {
            MemoryStorage.current.countryCode = countryCode
        }
        
        if let shopperLocale = config?["shopperLocale"] as? String {
            MemoryStorage.current.shopperLocale = shopperLocale
        }
        
        if let additionalData = config?["additionalData"] as? NSDictionary {
            if let allow3DS2 = additionalData["allow3DS2"] as? Bool {
                MemoryStorage.current.allow3DS2 = allow3DS2
            }
            if let executeThreeD = additionalData["executeThreeD"] as? Bool {
                MemoryStorage.current.executeThreeD = executeThreeD
            }
        }
        
        if let merchantAccount = config?["merchantAccount"] as? String {
            MemoryStorage.current.merchantAccount = merchantAccount
        }
        
        if let returnUrl = config?["returnUrl"] as? String {
            MemoryStorage.current.returnUrl = returnUrl
        }
        
        let environment = ConfigurationParser.getEnvironment(environmentName)
        let adyenApiContext = APIContext(environment: environment, clientKey: clientKey)
        self.dropInConfiguration = ConfigurationParser(adyenApiContext).parse(config!)
    }
    
    @objc
    func setModuleConfig(_ config: NSDictionary?) {
        print("Called setModuleConfig")
        
        guard let baseUrl = config?["baseUrl"] as? String else {
            reject("setModuleConfig missing baseUrl")
            return
        }
        
        MemoryStorage.current.baseUrl = URL(string: baseUrl)!
        
        if let debug = config?["debug"] as? Bool {
            MemoryStorage.current.debug = debug
            AdyenLogging.isEnabled = debug
        }
        
        if let headers = config?["headers"] as? [String: String] {
            MemoryStorage.current.headers = headers
        }
        
        if let endpoints = config?["endpoints"] as? NSDictionary {
            if let makePaymentEndpoint = endpoints["makePayment"] as? String {
                MemoryStorage.current.makePaymentEndpoint = makePaymentEndpoint
            }
            
            if let makeDetailsCallEndpoint = endpoints["makeDetailsCall"] as? String {
                MemoryStorage.current.makeDetailsCallEndpoint = makeDetailsCallEndpoint
            }
        }
    }
    
    @objc
    func setSubmitCallback(_ onSubmit: @escaping RCTResponseSenderBlock) {
        print("Called setSubmitCallback")
        self.onSubmitCallback = onSubmit
    }
    
    @objc
    func setAdditionalDetailsCallback(_ onAdditionalDetails: @escaping RCTResponseSenderBlock) {
        print("Called setAdditionalDetailsCallback")
        self.onAdditionalDetailsCallback = onAdditionalDetails
    }
    
    @objc
    func setPaymentResponse(_ paymentResponse: NSDictionary?) {
        print("Called setPaymentResponse")
        
        guard paymentResponse != nil else {
            return
        }

        do {
            let response = try AdyenDropInModule.decodeResponse(paymentResponse!)
            self.handleAsyncResponse(response)
        } catch let err {
            print(err.localizedDescription)
        }
    }
    
    @objc
    func setDetailsResponse(_ detailsResponse: NSDictionary?) {
        print("Called setDetailsResponse")
        
        guard detailsResponse != nil else {
            return
        }

        do {
            let response = try AdyenDropInModule.decodeResponse(detailsResponse!)
            self.handleAsyncResponse(response)
        } catch let err {
            print(err.localizedDescription)
        }
    }
    
    static func decodeResponse(_ response: NSDictionary) throws -> PaymentResponse {
        let JSON = try JSONSerialization.data(withJSONObject: response)
        let decoded = try Coder.decode(JSON) as PaymentResponse
        return decoded
    }
    
    func handleAsyncResponse(_ result: PaymentResponse) {
        switch result.resultCode {
        case .authorised, .pending, .received, .challengeShopper, .identifyShopper, .presentToShopper, .redirectShopper:
            if let action = result.action {
                handle(action)
            } else {
                DispatchQueue.main.async {
                    self.finish(with: result.resultCode)
                }
            }

        case .cancelled, .error, .refused:
            DispatchQueue.main.async {
                self.finish(with: result.resultCode)
            }
        }
    }
    
    @objc
    func start(_ paymentMethodsResponse: NSDictionary, resolveCallback: @escaping RCTResponseSenderBlock, rejectCallback: @escaping RCTResponseSenderBlock) {
        print("Called start")
        
        if (dropInConfiguration == nil) {
            reject("start was called without dropInConfig being set")
            return
        }
        
        guard let baseUrl = MemoryStorage.current.baseUrl,
              let clientKey = MemoryStorage.current.clientKey
        else {
            reject("Missing baseUrl or clientKey")
            return
        }
        
        let environment = Environment(baseURL: baseUrl)
        let apiContext = APIContext(environment: environment, clientKey: clientKey)
        self.apiClient = DefaultAPIClient(apiContext: apiContext)
        
        do {
            self.resolveCallback = resolveCallback
            self.rejectCallback = rejectCallback
            
            let jsonObject = try JSONSerialization.data(withJSONObject: paymentMethodsResponse)
            let paymentMethods = try Coder.decode(jsonObject) as PaymentMethods
            DispatchQueue.main.async {
                self.presenter = UIApplication.shared.delegate?.window??.rootViewController
                self.presentDropInComponent(paymentMethods)
            }
        } catch let err {
            reject("An error occurred while attempting to start payment: \(err.localizedDescription)")
        }
    }
    
    // MARK: - Adyen DropIn
    
    internal func presentDropInComponent(_ paymentMethods: PaymentMethods) {
        guard let dropInConfiguration = dropInConfiguration else { return }
        
        let dropInComponentStyle = DropInComponent.Style()
        let component = DropInComponent(paymentMethods: paymentMethods,
                                        configuration: dropInConfiguration,
                                        style: dropInComponentStyle,
                                        title: nil)
        component.delegate = self
        currentComponent = component
        
        presenter?.present(component.viewController, animated: true, completion: nil)
    }
    
    internal func finish(with response: PaymentsResponse) {
        let success = response.resultCode == .authorised || response.resultCode == .received || response.resultCode == .pending
        currentComponent?.finalizeIfNeeded(with: success)

        presenter?.dismiss(animated: true) { [weak self] in
            print("Dismiss successfully")
            
            do {
                let jsonObject = try JSONEncoder().encode(response)
                let str = String(data: jsonObject, encoding: .utf8)
                self?.resolve(str as Any)
            } catch let err {
                print(err.localizedDescription)
            }
        }
    }
    
    internal func finish(with resultCode: ResultCode) {
        let success = resultCode == .authorised || resultCode == .received || resultCode == .pending
        currentComponent?.finalizeIfNeeded(with: success)

        presenter?.dismiss(animated: true) { [weak self] in
            print("Dismiss successfully")
            
            do {
                let jsonObject = try JSONEncoder().encode(["resultCode": resultCode.rawValue])
                let str = String(data: jsonObject, encoding: .utf8)
                self?.resolve(str as Any)
            } catch let err {
                print(err.localizedDescription)
            }
        }
    }

    internal func finish(with error: Error) {
        currentComponent?.finalizeIfNeeded(with: false)

        presenter?.dismiss(animated: true) { [weak self] in
            if let componentError = error as? ComponentError {
                switch componentError {
                case .cancelled:
                    self?.reject("Cancelled")
                    break
                case .paymentMethodNotSupported:
                    self?.reject("Payment method not supported")
                    break
                }
            } else if let redirectComponentError = error as? RedirectComponent.Error {
                switch redirectComponentError {
                case .appNotFound:
                    self?.reject("App for payment method not found")
                    break
                }
            } else if let paymentsErrorResponse = error as? PaymentsErrorResponse {
                if (paymentsErrorResponse.resultCode == nil) {
                    // Try to create JSON object for error
                    if let jsonObject = try? JSONEncoder().encode(paymentsErrorResponse) {
                        let str = String(data: jsonObject, encoding: .utf8)
                        self?.reject(str as Any)
                    } else {
                    // If couldn't create JSON object, send just the message string
                        self?.reject((paymentsErrorResponse.message ?? paymentsErrorResponse.refusalReason ?? error.localizedDescription) as Any)
                    }
                // Finished with refused etc
                } else if let jsonObject = try? JSONEncoder().encode(paymentsErrorResponse) {
                    let str = String(data: jsonObject, encoding: .utf8)
                    self?.resolve(str as Any)
                } else {
                    self?.reject(paymentsErrorResponse.message as Any)
                }
            } else {
                self?.reject(error.localizedDescription)
            }
        }
    }
    
    // MARK: - Payment response handling
    private func paymentResponseHandler(result: Result<PaymentsResponse, Error>) {
        switch result {
        case let .success(response):
            if let action = response.action {
                handle(action)
            } else {
                finish(with: response)
            }
        case let .failure(error):
            finish(with: error)
        }
    }

    private func handle(_ action: Action) {
        (currentComponent as? DropInComponent)?.handle(action)
    }
}

extension AdyenDropInModule: DropInComponentDelegate {

    internal func didSubmit(_ data: PaymentComponentData, for paymentMethod: PaymentMethod, from component: DropInComponent) {
        if self.onSubmitCallback != nil {
            self.onSubmitCallback?([[
                "paymentMethod": data.paymentMethod.encodable.dictionary as Any,
                "storePaymentMethod": data.storePaymentMethod,
                "browserInfo": data.browserInfo as Any,
                "channel": "iOS"
            ]])
        } else {
            print("User did start: \(paymentMethod.name)")
            let headers = MemoryStorage.current.headers
            let path = MemoryStorage.current.makePaymentEndpoint
            let request = PaymentsRequest(headers: headers, path: path, data: data)
            apiClient?.perform(request, completionHandler: paymentResponseHandler)
        }
    }

    internal func didProvide(_ data: ActionComponentData, from component: DropInComponent) {
        if self.onAdditionalDetailsCallback != nil {
            self.onAdditionalDetailsCallback?([[
                "details": data.details.dictionary as Any,
                "paymentData": (data.paymentData ?? "") as String
            ]])
        } else {
            let headers = MemoryStorage.current.headers
            let path = MemoryStorage.current.makeDetailsCallEndpoint
            
            let request = PaymentDetailsRequest(
                headers: headers,
                path: path,
                details: data.details,
                paymentData: data.paymentData,
                merchantAccount: MemoryStorage.current.merchantAccount
            )
            apiClient?.perform(request, completionHandler: paymentResponseHandler)
        }
    }

    internal func didComplete(from component: DropInComponent) {
        finish(with: .authorised)
    }

    internal func didFail(with error: Error, from component: DropInComponent) {
        finish(with: error)
    }

    internal func didCancel(component: PaymentComponent, from dropInComponent: DropInComponent) {
        // Handle the event when the user closes a PresentableComponent.
        print("User did close: \(component.paymentMethod.name)")
    }

}

extension Encodable {

    var dictionary: [String: Any]? {
        guard let data = try? JSONEncoder().encode(self) else { return nil }
        return (try? JSONSerialization.jsonObject(with: data, options: .allowFragments)).flatMap { $0 as? [String: Any] }
    }

}

struct PaymentResponse: Decodable {

    internal let resultCode: ResultCode

    internal let action: Action?

    internal init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.resultCode = try container.decode(ResultCode.self, forKey: .resultCode)
        self.action = try container.decodeIfPresent(Action.self, forKey: .action)
    }

    private enum CodingKeys: String, CodingKey {
        case resultCode
        case action
    }

}

File: android/src/main/java/com/reactnativeadyendropin/AdyenDropInModule.kt
package com.reactnativeadyendropin

import android.app.Activity
import android.content.Intent
import android.util.Log
import android.util.Log.VERBOSE
import com.adyen.checkout.components.model.PaymentMethodsApiResponse
import com.adyen.checkout.core.log.Logger
import com.adyen.checkout.core.model.getStringOrNull
import com.adyen.checkout.dropin.DropIn
import com.adyen.checkout.dropin.DropInConfiguration
import com.adyen.checkout.dropin.DropInResult
import com.adyen.checkout.dropin.service.DropInServiceResult
import com.facebook.react.bridge.*
import com.reactnativeadyendropin.data.storage.MemoryStorage
import com.reactnativeadyendropin.service.AdyenDropInService
import org.json.JSONObject
import org.koin.core.component.KoinComponent
import org.koin.core.component.get

class AdyenDropInModule(private val reactContext : ReactApplicationContext): ReactContextBaseJavaModule(reactContext), KoinComponent, ActivityEventListener  {
  private val TAG = "AdyenDropInModule"

  private val memoryStorage: MemoryStorage = get()

  private var dropInConfiguration: DropInConfiguration? = null

  private var resolveCallback: Callback? = null

  private var rejectCallback: Callback? = null

  init {
    reactContext.addActivityEventListener(this)
  }

  override fun getName(): String {
    return "AdyenDropInModule"
  }

  @ReactMethod
  fun setDropInConfig(config: ReadableMap?) {
    Log.d(TAG, "Called setDropInConfig")

    if (config == null) {
      rejectCallback?.invoke("setDropInConfig was called without a config object")
    } else {
      try {
        val clientKey = config.getString("clientKey")
        val parser = ConfigurationParser(clientKey!!, this.reactContext)
        this.dropInConfiguration = parser.parse(config)

        if (config.hasKey("shopperReference")) {
          memoryStorage.shopperReference = config.getString("shopperReference")!!
        }

        if (config.hasKey("amount")) {
          val amount = config.getMap("amount")!!
          memoryStorage.amountValue = amount.getInt("value")
          memoryStorage.amountCurrency = amount.getString("currencyCode")!!
        }

        if (config.hasKey("countryCode")) {
          memoryStorage.countryCode = config.getString("countryCode")!!
        }

        if (config.hasKey("shopperLocale")) {
          memoryStorage.shopperLocale = config.getString("shopperLocale")!!
        }

        if (config.hasKey("additionalData")) {
          val additionalData = config.getMap("additionalData")!!
          memoryStorage.allow3DS2 = additionalData.getBoolean("allow3DS2")
          memoryStorage.executeThreeD = additionalData.getBoolean("executeThreeD")
        }
      } catch (err: Error) {
        Log.d(TAG, "Error in setDropInConfiguration - ${err.message}")
        Log.e(TAG, err.toString())
        rejectCallback?.invoke(err.message)
      }
    }
  }

  @ReactMethod
  fun setModuleConfig(config: ReadableMap?) {
    Log.d(TAG, "Called setModuleConfig")

    if (config == null) {
      rejectCallback?.invoke("setModuleConfig was called without a config object")
    } else {
      try {
        if (config.hasKey("baseUrl")) {
          memoryStorage.baseUrl = config.getString("baseUrl")!!
        }

        if (config.hasKey("debug")) {
          val debug = config.getBoolean("debug")
          if (debug is Boolean) {
            memoryStorage.debug = debug
            Logger.setLogcatLevel(VERBOSE)
          }
        }

        if (config.hasKey("disableNativeRequests")) {
          memoryStorage.disableNativeRequests = config.getBoolean("disableNativeRequests")
        }

        if (config.hasKey("headers")) {
          val map = config.getMap("headers")!!
          memoryStorage.headers = RNUtils.readableMapToStringMap(map)
        }

        if (config.hasKey("queryParameters")) {
          val map = config.getMap("queryParameters")!!
          memoryStorage.queryParameters = RNUtils.readableMapToStringMap(map)
        }

        if (config.hasKey("endpoints")) {
          val endpoints = config.getMap("endpoints")!!
          memoryStorage.makePaymentEndpoint = endpoints.getString("makePayment")!!
          memoryStorage.makeDetailsCallEndpoint = endpoints.getString("makeDetailsCall")!!
        }
      } catch (err: Error) {
        Log.d(TAG, "Error in setModuleConfig - ${err.message}")
        Log.e(TAG, err.toString())
        rejectCallback?.invoke("Failed to set module config. Check the config object values and that its types are correct. ${err.message}")
      }
    }
  }

  @ReactMethod
  fun setSubmitCallback(callback: Callback?) {
    Log.d(TAG, "Called setSubmitCallback")
    memoryStorage.onSubmitCallback = callback
  }

  @ReactMethod
  fun setAdditionalDetailsCallback(callback: Callback?) {
    Log.d(TAG, "Called setAdditionalDetailsCallback")
    memoryStorage.onAdditionalDetailsCallback = callback
  }

  @ReactMethod
  fun setPaymentResponse(paymentResponse: ReadableMap?) {
    Log.d(TAG, "Called setPaymentResponse")
    if (paymentResponse == null) return

    try {
      AdyenDropInService.handleAsyncResponse(paymentResponse)
    } catch (err: Exception) {
      this.handleError(err.message  ?: "Unknown error in setPaymentResponse")
    }
  }

  @ReactMethod
  fun setDetailsResponse(detailsResponse: ReadableMap?) {
    Log.d(TAG, "Called setDetailsResponse")
    if (detailsResponse == null) return

    try {
      AdyenDropInService.handleAsyncResponse(detailsResponse)
    } catch (err: Exception) {
      this.handleError(err.message  ?: "Unknown error in setDetailsResponse")
    }
  }

  @ReactMethod
  fun start(paymentMethodsResponse: ReadableMap, resolveCallback: Callback, rejectCallback: Callback) {
    Log.d(TAG, "Called start")

    if (this.dropInConfiguration == null) {
      rejectCallback.invoke("start was called without dropInConfig being set")
      return
    }

    try {
      this.resolveCallback = resolveCallback
      this.rejectCallback = rejectCallback

      val jsonObject = RNUtils.convertMapToJson(paymentMethodsResponse)
      val paymentMethodsApiResponse = PaymentMethodsApiResponse.SERIALIZER.deserialize(jsonObject)
      DropIn.startPayment(this.reactContext.currentActivity!!, paymentMethodsApiResponse, this.dropInConfiguration!!)
    } catch (err: Error) {
      rejectCallback.invoke("An error occurred while attempting to start payment: ${err.message}")
    }
  }

  fun handleError(reason: String?) {
    Log.d(TAG, "handleError - ${reason}")
    this.rejectCallback?.invoke(reason)
  }

  fun handleCancelled() {
    Log.d(TAG, "handleCancelled")
    this.rejectCallback?.invoke("Cancelled")
  }

  fun handleFinished(result: String) {
    Log.d(TAG, "handleFinished - ${result}")
    this.resolveCallback?.invoke(result)
  }

  override fun onActivityResult(activity: Activity?, requestCode: Int, resultCode: Int, data: Intent?) {
    Log.d(TAG, "onActivityResult - ${activity}, ${requestCode}, ${resultCode}, ${data}")

    val dropInResult = DropIn.handleActivityResult(requestCode, resultCode, data) ?: return
    when (dropInResult) {
      is DropInResult.Error -> handleError(dropInResult.reason)
      is DropInResult.CancelledByUser -> handleCancelled()
      is DropInResult.Finished -> {
        Log.d(TAG, "dropInResult - ${dropInResult.result}")

        try {
          val json = JSONObject(dropInResult.result)
          if (json.getStringOrNull("resultCode") == null) {
            handleError(dropInResult.result ?: json.getStringOrNull("message") ?: json.getStringOrNull("refusalReason") ?: "Unknown error")
          } else {
             handleFinished(dropInResult.result)
          }
        } catch (err: Error) {
          Log.e(TAG, err.message ?: "Unknown parsing error")
          handleFinished(dropInResult.result)
        }
      }
    }
  }

  override fun onNewIntent(intent: Intent?) {
    Log.d(TAG, "onNewIntent - ${intent}")
  }
}
